File: ems-backend/src/main/java/com/hrms/ems/service/DepartmentService.java
package com.hrms.ems.service;

import com.hrms.ems.model.Department;
import com.hrms.ems.repository.DepartmentRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class DepartmentService {

    @Autowired
    private DepartmentRepository departmentRepository;

    public List<Department> getAllDepartments() {
        return departmentRepository.findAll();
    }

    public Optional<Department> getDepartmentById(@org.springframework.lang.NonNull Long id) {
        return departmentRepository.findById(id);
    }

    public Department createDepartment(@org.springframework.lang.NonNull Department department) {
        return departmentRepository.save(department);
    }

    public Department updateDepartment(@org.springframework.lang.NonNull Long id, Department departmentDetails) {
        Department department = departmentRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Department not found"));
        
        department.setName(departmentDetails.getName());
        department.setDescription(departmentDetails.getDescription());
        
        return departmentRepository.save(department);
    }

    public void deleteDepartment(@org.springframework.lang.NonNull Long id) {
        departmentRepository.deleteById(id);
    }
}

File: ems-frontend/src/services/department.service.js
import api from './api';

const getAllDepartments = async () => {
    const response = await api.get('/departments');
    return response.data;
};

const createDepartment = async (departmentData) => {
    const response = await api.post('/departments', departmentData);
    return response.data;
};

const updateDepartment = async (id, departmentData) => {
    const response = await api.put(`/departments/${id}`, departmentData);
    return response.data;
};

const deleteDepartment = async (id) => {
    const response = await api.delete(`/departments/${id}`);
    return response.data;
};

const departmentService = {
    getAllDepartments,
    createDepartment,
    updateDepartment,
    deleteDepartment,
};

export default departmentService;
